src/jackdaw/streams/extras.clj

```diff
@@ -3,13 +3,117 @@
   {:license "BSD 3-Clause License <https://github.com/FundingCircle/jackdaw/blob/master/LICENSE>"}
   (:require [clojure.tools.logging :as log]
             [jackdaw.streams :as js]
+            [jackdaw.streams.interop :as jsi]
+            [jackdaw.streams.protocols :as jsp]
             [clojure.spec.alpha :as s])
-  (:import org.apache.kafka.common.TopicPartition
-           org.apache.kafka.streams.processor.StateRestoreListener
+  (:import (jackdaw.streams.interop CljKStream CljKTable CljGlobalKTable CljKGroupedTable
+                                    CljKGroupedStream CljTimeWindowedKStream CljSessionWindowedKStream)
+           org.apache.kafka.common.TopicPartition
+           (org.apache.kafka.streams StreamsBuilder)
+           (org.apache.kafka.streams.kstream KStream KTable KGroupedStream KGroupedTable
+                                             GlobalKTable TimeWindowedKStream
+                                             SessionWindowedKStream Transformer ValueTransformer)
+           (org.apache.kafka.streams.processor ProcessorContext StateRestoreListener)
+           (org.apache.kafka.streams.state Stores)
            [java.time Instant Duration]))
 
 (set! *warn-on-reflection* true)
 
+(defn transformer-with-ctx
+   "Helper to create a Transformer for use inside the jackdaw transform wrapper.
+   Passed function should take three args - the context, key and value for the stream.
+   The processor context allows access to stream internals such as state stores.
+   Result is returned from the transform. E.g.
+   ```
+   (-> builder
+       (k/stream topic)
+       (k/transform
+         (kx/transformer-with-ctx
+           (fn [ctx k v]
+             ...))))
+   ```"
+   [xfm]
+   (fn []
+     (let [ctx (atom nil)]
+       (reify Transformer
+         (init [_ processor-context]
+           (reset! ctx processor-context))
+         (close [_])
+         (transform [_ k v]
+           (xfm @ctx k v))))))
+
+(defn value-transformer-with-ctx
+   "Helper to create a ValueTransformer for use inside the jackdaw transform-values wrapper.
+   Passed function should take two args - the context and value for the stream.
+   The processor context allows access to stream internals such as state stores.
+   Result is returned from the transform-values. E.g.
+   ```
+   (-> builder
+       (k/stream topic)
+       (k/transform-values
+         (kx/value-transformer-with-ctx
+           (fn [ctx v]
+             ...))))
+   ```"
+  [xfm]
+  (fn []
+    (let [ctx (atom nil)]
+      (reify ValueTransformer
+        (init [_ processor-context]
+          (reset! ctx processor-context))
+        (close [_])
+        (transform [_ v]
+          (xfm @ctx v))))))
+
+(defn with-state-store
```

Comment:
something like `with-kv-state-store` or mentioning in the docstring would be more informative since it's not the only type of state store

Reply:
👍